lib/ansible/modules/packaging/os/snap.py

```diff
@@ -0,0 +1,193 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+# Copyright: (c) 2018, Stanislas Lange (angristan) <angristan@pm.me>
+# Copyright: (c) 2018, Victor Carceler <vcarceler@iespuigcastellar.xeill.net>
+
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+ANSIBLE_METADATA = {
+    'metadata_version': '1.1',
+    'status': ['preview'],
+    'supported_by': 'community'
+}
+
+DOCUMENTATION = '''
+---
+module: snap
+
+short_description: Manages snaps
+
+version_added: "2.8"
+
+description:
+    - "Manages snaps packages."
+
+options:
+    name:
+        description:
+            - Name of the snap to install or remove.
+        required: true
+    state:
+        description:
+            - Desired state of the package.
+        required: false
+        default: present
+        choices: [ absent, present ]
+    classic:
+        description:
+            - Confinement policy. The classic confinment allows a snap to have
+              the same level of access to the system as "classic" packages,
+              like those managed by APT. This option corresponds to the --classic argument.
+        type: bool
+        required: false
+        default: False
+
+author:
+    - Victor Carceler (vcarceler@iespuigcastellar.xeill.net)
+    - Stanislas Lange (angristan) <angristan@pm.me>
+'''
+
+EXAMPLES = '''
+# Install "foo" snap
+- name: Install foo
+  snap:
+    name: foo
+
+# Remove "foo" snap
+- name: Remove foo
+  snap:
+    name: foo
+    state: absent
+
+# Install a snap with classic confinement
+- name: Install "foo" with option --classic
+  snap:
+    name: foo
+    classic: yes
+'''
+
+RETURN = '''
+classic:
+    description: Whether or not the snaps were installed with the classic confinement
+    type: boolean
+    returned: When snaps are installed
+cmd:
+    description: The command that was executed on the host
+    type: string
+    returned: When changed is true
+'''
+
+from ansible.module_utils.basic import AnsibleModule
+import re
+
+
+def snap_exists(module, snap_name):
+    snap_path = module.get_bin_path("snap", True)
+    cmd_parts = [snap_path, 'info', snap_name]
+    cmd = ' '.join(cmd_parts)
+    rc, out, err = module.run_command(cmd, check_rc=False)
+
+    return rc == 0
+
+
+def is_snap_installed(module, snap_name):
+    snap_path = module.get_bin_path("snap", True)
+    cmd_parts = [snap_path, 'list', snap_name]
+    cmd = ' '.join(cmd_parts)
+    rc, out, err = module.run_command(cmd, check_rc=False)
+
+    return rc == 0
+
+
+def install_snaps(module, snap_names):
+    snaps_not_installed = list()
+    for snap_name in snap_names:
+        if is_snap_installed(module, snap_name):
+            continue
+        snaps_not_installed.append(snap_name)
+
+    if not snaps_not_installed:
+        module.exit_json(classic=module.params['classic'], changed=False)
```

Comment:
you could hold some default return kwargs and then modify them, unpack when exiting:
```python
exit_kwargs = {
    'classic': module.params['classic'],
    'changed': False,
}
...
if ...:
    module.exit_json(**exit_kwargs)
...
... # after some time
...
exit_kwargs['changed'] = True
...
... # other exits:
module.exit_json(cmd=..., stdout=..., stderr=..., **exit_kwargs)  # same for fail_
```

Reply:
Here's a diff:

![screen shot 2018-09-15 at 01 46 57](https://user-images.githubusercontent.com/11699655/45579616-4c018b00-b889-11e8-8ed0-5270564e2f0b.png)

Do you think this is better?

Reply:
Looks nicer :)

Reply:
Done. I tried to add it for `remove_snap`, but it just seems inefficient